neqo-transport/src/connection.rs

```diff
@@ -3763,32 +3770,51 @@ mod tests {
         assert_eq!(pkt0.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
     }
 
-    // Handle sending a bunch of bytes from one connection to another, until
-    // something stops us from sending.
-    fn send_bytes(src: &mut Connection, stream: u64, now: Instant) -> Vec<Datagram> {
+    /// This fills the congestion window from a single source.
+    /// As the pacer will interfere with this, this moves time forward
+    /// as `Output::Callback` is received.  Because it is hard to tell
+    /// from the return value whether a timeout is to ACK delay, PTO, or
+    /// pacing, this looks at the congestion window to tell when to stop.
+    /// Returns a list of datagrams and the new time.
+    fn fill_cwnd(src: &mut Connection, stream: u64, mut now: Instant) -> (Vec<Datagram>, Instant) {
+        const BLOCK_SIZE: usize = 4_096;
         let mut total_dgrams = Vec::new();
 
+        qtrace!(
+            "fill_cwnd starting cwnd: {}",
+            src.loss_recovery.cwnd_avail()
+        );
+
         loop {
-            let bytes_sent = src.stream_send(stream, &[0x42; 4_096]).unwrap();
-            qtrace!("send_bytes wrote {} bytes", bytes_sent);
-            if bytes_sent == 0 {
+            let bytes_sent = src.stream_send(stream, &[0x42; BLOCK_SIZE]).unwrap();
+            qtrace!("fill_cwnd wrote {} bytes", bytes_sent);
+            if bytes_sent < BLOCK_SIZE {
                 break;
             }
         }
 
         loop {
             let pkt = src.process_output(now);
-            qtrace!("send_bytes output: {:?}", pkt);
+            qtrace!(
+                "fill_cwnd cwnd remaining={}, output: {:?}",
+                src.loss_recovery.cwnd_avail(),
+                pkt
+            );
             match pkt {
                 Output::Datagram(dgram) => {
                     total_dgrams.push(dgram);
                 }
-                Output::Callback(_) => break,
+                Output::Callback(t) => {
+                    if src.loss_recovery.cwnd_avail() < ACK_ONLY_SIZE_LIMIT {
```

Comment:
This is where we need special info, which includes when we stop sending packets.  This is unfortunate, but I don't see how this can be avoided without creating a new API.  Maybe that API is something we want, but I didn't want to create one without first discussing it and having a use case for it that isn't just making our tests work.